pkg/kubectl/cmd/drain.go

```diff
@@ -195,148 +212,149 @@ func (o *DrainOptions) RunDrain() error {
+	case "ReplicaSet":
+		return o.client.ExtensionsClient.ReplicaSets(sr.Reference.Namespace).Get(sr.Reference.Name)
+	}
+	return nil, fmt.Errorf("Unknown controller kind %q", sr.Reference.Kind)
+}
+
+func (o *DrainOptions) getPodCreator(pod api.Pod) (*api.SerializedReference, error) {
+	creatorRef, found := pod.ObjectMeta.Annotations[controller.CreatedByAnnotation]
+	if !found {
+		return nil, nil
+	}
+
+	// Now verify that the specified creator actually exists.
+	sr := &api.SerializedReference{}
+	if err := runtime.DecodeInto(o.factory.Decoder(true), []byte(creatorRef), sr); err != nil {
+		return nil, err
+	}
+	// We assume the only reason for an error is because the controller is
+	// gone/missing, not for any other cause.  TODO(mml): something more
+	// sophisticated than this
+	_, err := o.getController(sr)
 	if err != nil {
-		return []api.Pod{}, err
+		return nil, err
 	}
+	return sr, nil
+}
 
-	daemonSetErrors := !o.IgnoreDaemonsets && len(daemonSetPodNames) > 0
-	unreplicatedErrors := !o.Force && len(unreplicatedPodNames) > 0
+func (o *DrainOptions) unreplicatedFilter(pod api.Pod) (bool, *warning, *fatal) {
+	sr, err := o.getPodCreator(pod)
+	if err != nil {
+		return false, nil, &fatal{err.Error()}
+	}
+	if sr != nil {
+		return true, nil, nil
+	}
+	if !o.Force {
+		return false, nil, &fatal{kUnmanagedFatal}
+	}
+	return true, &warning{kUnmanagedWarning}, nil
+}
 
-	switch {
-	case daemonSetErrors && unreplicatedErrors:
-		return []api.Pod{}, errors.New(unmanagedMsg(unreplicatedPodNames, daemonSetPodNames, true))
-	case daemonSetErrors && !unreplicatedErrors:
-		return []api.Pod{}, errors.New(unmanagedMsg([]string{}, daemonSetPodNames, true))
-	case unreplicatedErrors && !daemonSetErrors:
-		return []api.Pod{}, errors.New(unmanagedMsg(unreplicatedPodNames, []string{}, true))
+func (o *DrainOptions) daemonsetFilter(pod api.Pod) (bool, *warning, *fatal) {
+	sr, err := o.getPodCreator(pod)
+	if err != nil {
+		return false, nil, &fatal{err.Error()}
+	}
+	if sr == nil || sr.Reference.Kind != "DaemonSet" {
+		return true, nil, nil
+	}
+	if _, err := o.client.DaemonSets(sr.Reference.Namespace).Get(sr.Reference.Name); err != nil {
+		return false, nil, &fatal{err.Error()}
 	}
+	if !o.IgnoreDaemonsets {
+		return false, nil, &fatal{kDaemonsetFatal}
+	}
+	return false, &warning{kDaemonsetWarning}, nil
+}
 
-	if len(unreplicatedPodNames) > 0 {
-		fmt.Fprintf(o.out, "WARNING: About to delete these %s\n", unmanagedMsg(unreplicatedPodNames, []string{}, false))
+func mirrorPodFilter(pod api.Pod) (bool, *warning, *fatal) {
+	if _, found := pod.ObjectMeta.Annotations[types.ConfigMirrorAnnotationKey]; found {
+		return false, nil, nil
 	}
-	if len(daemonSetPodNames) > 0 {
-		fmt.Fprintf(o.out, "WARNING: Skipping %s\n", unmanagedMsg([]string{}, daemonSetPodNames, false))
+	return true, nil, nil
+}
+
+func hasLocalStorage(pod api.Pod) bool {
+	for _, volume := range pod.Spec.Volumes {
+		if volume.HostPath != nil || volume.EmptyDir != nil {
```

Comment:
HostDir will survive the drain... depending on what happens next.  If the intent is just a reboot and HostDir is durable storage, then yes.  If the intent is repairs, or upgrade of a VM where that upgrade is implemented by deleting the host first, then no.  If HostDir is in `/tmp` or a similar directory also no.

Depends on exactly which scenarios we're trying to prevent here.